src/PowerShell/PowerShellManager.cs

```diff
@@ -50,8 +50,65 @@ internal PowerShellManager(ILogger logger)
-            // TODO: The Azure Functions Host might supply these differently. This might change but works for the demo
+            // Check if Az.Profile is available
+            Collection<PSModuleInfo> azprofile = _pwsh.AddCommand("Get-Module")
+                .AddParameter("ListAvailable")
+                .AddParameter("Name", "Az.Profile")
+                .InvokeAndClearCommands<PSModuleInfo>();
+
+            if (azprofile.Count == 0)
+            {
+                _logger.Log(LogLevel.Warning, "Required module to automatically authenticate with Azure `Az.Profile` was not found in the PSModulePath.");
+                return;
+            }
+
+            // Try to authenticate to Azure using MSI
+            string msiSecret = Environment.GetEnvironmentVariable("MSI_SECRET");
+            string msiEndpoint = Environment.GetEnvironmentVariable("MSI_ENDPOINT");
+            string accountId = Environment.GetEnvironmentVariable("WEBSITE_SITE_NAME");
+
+            if (!string.IsNullOrEmpty(msiSecret) &&
+                !string.IsNullOrEmpty(msiEndpoint) &&
+                !string.IsNullOrEmpty(accountId))
+            {
+                // NOTE: There is a limitation in Azure PowerShell that prevents us from using the parameter set:
+                // Connect-AzAccount -MSI or Connect-AzAccount -Identity
+                // see this GitHub issue https://github.com/Azure/azure-powershell/issues/7876
+                // As a workaround, we can all an API endpoint on the MSI_ENDPOINT to get an AccessToken and use that to authenticate
+                Collection<PSObject> response = _pwsh.AddCommand("Microsoft.PowerShell.Utility\\Invoke-RestMethod")
+                    .AddParameter("Method", "Get")
+                    .AddParameter("Headers", new Hashtable {{ "Secret", msiSecret }})
+                    .AddParameter("Uri", $"{msiEndpoint}?resource=https://management.azure.com&api-version=2017-09-01")
+                    .InvokeAndClearCommands<PSObject>();
+
+                if(_pwsh.HadErrors) 
+                {
+                    _logger.Log(LogLevel.Trace, "Failed to Authenticate to Azure via MSI. Check the logs for the errors generated.");
+                }
+                else
+                {
+                    // We have successfully authenticated to Azure so we can return out.
+                    using (ExecutionTimer.Start(_logger, "Authentication to Azure"))
+                    {
+                        _pwsh.AddCommand("Az.Profile\\Connect-AzAccount")
+                            .AddParameter("AccessToken", response[0].Properties["access_token"].Value)
+                            .AddParameter("AccountId", accountId)
+                            .InvokeAndClearCommands();
+
+                        if(_pwsh.HadErrors)
+                        {
+                            _logger.Log(LogLevel.Trace, "Failed to Authenticate to Azure. Check the logs for the errors generated.");
```

Comment:
Same comments as above. And also call out in the message that we failed with MSI, because we might succeed later when trying with service principal.

Reply:
Warning?

Also, they're generated from the worker handling Write-Error